packages/stream/common/ContractUtils.ts

```diff
@@ -0,0 +1,54 @@
+import BN from "bn.js";
+
+export const calculateUnlocked = (
+  depositedAmount: BN,
+  cliff: number,
+  cliffAmount: BN,
+  end: number,
+  currentTimestamp: number,
+  lastRateChangeTime: number,
+  period: number,
+  amountPerPeriod: BN,
+  fundsUnlockedAtLastRateChange: BN
+): BN => {
+  const deposited = depositedAmount;
+
+  if (currentTimestamp < cliff) return new BN(0);
+  if (currentTimestamp > end) return deposited;
+
+  const streamed =
+    lastRateChangeTime !== 0
+      ? nonLinearReleaseRateUnlocked(
+          currentTimestamp,
+          lastRateChangeTime,
+          period,
+          amountPerPeriod,
+          fundsUnlockedAtLastRateChange
+        )
+      : linearReleaseRateUnlocked(currentTimestamp, cliff, period, cliffAmount, amountPerPeriod);
+
+  return streamed.lt(deposited) ? streamed : deposited;
+};
+
+const linearReleaseRateUnlocked = (
+  currentTime: number,
+  cliffTime: number,
+  period: number,
+  cliffAmount: BN,
+  releaseRate: BN
+) => {
+  const perPeriod = new BN(Math.floor((currentTime - cliffTime) / period));
+  return cliffAmount.add(perPeriod.mul(releaseRate));
+};
+
+const nonLinearReleaseRateUnlocked = (
```

Comment:
Technically speaking it's still a linear contract with updated rate. And those 2 functions looks oddly similar.

I think you can use either of those, for one you pass `cliffTime` and `cliffAmount`, for the other `lastRateChangeTime` and `fundsUnlockedAtLastRateChange`. The calculation remains the same from what I can see.

Not blocking in any way, feel free to ignore this one.